From 69d93a02c36c3c9bd49fee6da0c28272a0346cf2 Mon Sep 17 00:00:00 2001
From: Stefan Hallas Mulvad <kongen@greenwavereality.com>
Date: Wed, 13 Nov 2013 23:47:38 +0100
Subject: [PATCH 04/13] cs75xx gpio keys

Signed-off-by: Hoang Tran <hoang.tran@greenwavereality.com>
Signed-off-by: Hoang Tran <hoang.tran@greenwavesystems.com>
---
 drivers/input/keyboard/gpio_keys.c | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/drivers/input/keyboard/gpio_keys.c b/drivers/input/keyboard/gpio_keys.c
index a877a0e..52f524d 100644
--- a/drivers/input/keyboard/gpio_keys.c
+++ b/drivers/input/keyboard/gpio_keys.c
@@ -336,12 +336,6 @@ static void gpio_keys_gpio_report_event(struct gpio_button_data *bdata)
 			input_event(input, type, button->code, button->value);
 	} else {
 		input_event(input, type, button->code, !!state);
-#ifdef CONFIG_ARCH_GOLDENGATE
-		/* GPIO not support both edge interrupt, so we must send report
-		   an opposite edge-trigger(release) event manually */
-		input_sync(input);
-		input_event(input, type, button->code, !state);
-#endif
 	}
 	input_sync(input);
 }
@@ -364,6 +358,10 @@ static void gpio_keys_gpio_timer(unsigned long _data)
 static irqreturn_t gpio_keys_gpio_isr(int irq, void *dev_id)
 {
 	struct gpio_button_data *bdata = dev_id;
+#ifdef CONFIG_ARCH_GOLDENGATE
+	const struct gpio_keys_button *button = bdata->button;
+	unsigned long irqflags;
+#endif
 
 	BUG_ON(irq != bdata->irq);
 
@@ -373,6 +371,12 @@ static irqreturn_t gpio_keys_gpio_isr(int irq, void *dev_id)
 	else
 		schedule_work(&bdata->work);
 
+#ifdef CONFIG_ARCH_GOLDENGATE
+	/* toogle irqflag */
+	irqflags = gpio_get_value(button->gpio) ? IRQF_TRIGGER_FALLING : IRQF_TRIGGER_RISING;
+	irq_set_irq_type(bdata->irq, irqflags);
+#endif
+
 	return IRQ_HANDLED;
 }
 
@@ -479,7 +483,8 @@ static int __devinit gpio_keys_setup_key(struct platform_device *pdev,
 			    gpio_keys_gpio_timer, (unsigned long)bdata);
 
 		isr = gpio_keys_gpio_isr;
-#ifdef CONFIG_ARCH_GOLDENGATE /* GPIO not support both edge interrupt */
+#ifdef CONFIG_ARCH_GOLDENGATE
+		/* GPIO not support both edge interrupt */
 		irqflags = button->active_low ? IRQF_TRIGGER_FALLING : IRQF_TRIGGER_RISING;
 #else
 		irqflags = IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING;
-- 
2.5.0

