--- rtl8194_driver_v1.6_20151022.001/driver/rtl8192cd/8192cd_rx.c	2017-01-27 12:13:57.605356234 -0800
+++ rtl8194_driver_v1.6_20151022.001/driver/rtl8192cd/8192cd_rx.c	2017-01-27 12:16:47.767288996 -0800
@@ -2371,7 +2371,7 @@
 		setup_timer = 1;
 	}
 	else if (CIRC_SPACE(current_idx, priv->pshare->rc_timer_tail, RC_TIMER_NUM) == 0) {
-		DEBUG_ERR("%s: %s, RC timer overflow!\n", priv->dev->name, __FUNCTION__ );
+		DEBUG_WARN("%s: %s, RC timer overflow!\n", priv->dev->name, __FUNCTION__ );
 		return -1;
 	}
 	else {	// some items in timer queue
@@ -7151,7 +7151,7 @@
 	{
 		status = _RSON_CLS2_;
 		priv->ext_stats.rx_data_drops++;
-		DEBUG_ERR("RX DROP: class 2 error!\n");
+		DEBUG_WARN("RX DROP: class 2 error!\n");
 		goto data_defrag_error;
 	}
 	else if (!(pstat->state & WIFI_ASOC_STATE))
--- rtl8194_driver_v1.6_20151022.001/driver/rtl8192cd/8192cd_tx.c	2017-01-27 12:13:57.611356161 -0800
+++ rtl8194_driver_v1.6_20151022.001/driver/rtl8192cd/8192cd_tx.c	2017-01-27 12:16:06.522790061 -0800
@@ -1061,7 +1061,7 @@
 		(*((UINT16 *)((UINT8 *)skb->data + ETH_ALEN*2)) != htons(0x888e)) &&
 		(!(OPMODE & WIFI_ADHOC_STATE)))
 	{
-		DEBUG_ERR("TX DROP: control port not authorized!\n");
+		DEBUG_WARN("TX DROP: control port not authorized!\n");
 		rtl8192cd_tx_xmitSkbFail(priv, skb, dev, wdsDev, txcfg);
 		goto stop_proc;
 	}
